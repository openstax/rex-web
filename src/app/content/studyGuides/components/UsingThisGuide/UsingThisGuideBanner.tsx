import * as Cookies from 'js-cookie';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { Times } from 'styled-icons/fa-solid/Times';
import { useAnalyticsEvent } from '../../../../../helpers/analytics';
import { PlainButton } from '../../../../components/Button';
import { H2, h4MobileStyle } from '../../../../components/Typography/headings';
import theme from '../../../../theme';
import { disablePrint } from '../../../components/utils/disablePrint';
import { filters } from '../../../styles/PopupConstants';
import desktopBanner from './assets/banner.png';
import mobileBanner from './assets/banner_mobile.png';
import { cookieUTG } from './constants';

// tslint:disable-next-line: variable-name
const BannerWrapper = styled.div`
  outline: none;
  position: relative;
  padding: ${filters.dropdownToggle.topBottom.desktop}rem ${filters.dropdownToggle.sides.desktop}rem;
  width: 100%;
  background: ${theme.color.black};
  margin-bottom: 1rem;
  ${theme.breakpoints.mobileSmall(css`
    padding: ${filters.dropdownToggle.topBottom.mobile}rem ${filters.dropdownToggle.sides.mobile}rem;
  `)}
  ${disablePrint}
`;

// tslint:disable-next-line: variable-name
export const BannerImage = styled.img`
  width: 100%;
  padding: 0 4.2rem;
  ${theme.breakpoints.mobileMedium(css`
    max-width: 30rem;
    padding: 0;
  `)}
`;

// tslint:disable-next-line: variable-name
const UsingThisGuideTitle = styled(H2)`
  text-align: center;
  width: 100%;
  color: ${theme.color.white};
  ${theme.breakpoints.mobile(css`
    ${h4MobileStyle}
    color: ${theme.color.white};
  `)}
`;

// tslint:disable-next-line:variable-name
export const CloseIconButton = styled(PlainButton)`
  position: absolute;
  border-radius: 50%;
  background: ${theme.color.white};
  display: flex;
  align-items: center;
  justify-content: center;
  top: ${filters.dropdownToggle.topBottom.desktop}rem;
  right: ${filters.dropdownToggle.sides.desktop}rem;
  ${theme.breakpoints.mobileSmall(css`
    top: ${filters.dropdownToggle.topBottom.mobile}rem;
    right: ${filters.dropdownToggle.sides.mobile}rem;
  `)}
`;

// tslint:disable-next-line:variable-name
export const CloseIcon = styled(Times)`
  color: ${theme.color.black};
  height: 2.8rem;
  width: 2.8rem;
  padding: 0.4rem;
  ${theme.breakpoints.mobile(css`
    height: 1.6rem;
    width: 1.6rem;
  `)}
`;

// tslint:disable-next-line: variable-name
const HeaderWrapper = styled.div`
  display: flex;
  align-items: center;
`;

// tslint:disable-next-line: variable-name
const BodyWrapper = styled.div`
  display: flex;
  justify-content: center;
`;
interface Props {
  onClick: () => void;
  show: boolean;
}

// tslint:disable-next-line:variable-name
const UsingThisGuideBanner = (props: Props) => {
  const intl = useIntl();
  const trackOpenUTG = useAnalyticsEvent('openUTG');

  React.useEffect(() => {
    // Send GA event except when banner is open by default (cookie is set to 'true').
    const isCookieSet = Cookies.get(cookieUTG) === 'true';
    if (props.show && isCookieSet) {
      trackOpenUTG();
    } else if (props.show) {
      Cookies.set(cookieUTG, 'true', {expires: 365 * 20});
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.show]);

  if (!props.show) { return null; }

  return <BannerWrapper>
    <HeaderWrapper>
      <FormattedMessage id='i18n:studyguides:popup:using-this-guide'>
        {(msg) => <UsingThisGuideTitle>{msg}</UsingThisGuideTitle>}
      </FormattedMessage>
    </HeaderWrapper>
    <BodyWrapper>
      <picture>
        <source media={theme.breakpoints.mobileMediumQuery} srcSet={mobileBanner} />
        <BannerImage
          src={desktopBanner}
          alt={intl.formatMessage({id: 'i18n:studyguides:popup:using-this-guide:alt'})}
          tabIndex={0}
        />
      </picture>
    </BodyWrapper>
      <CloseIconButton
        onClick={props.onClick}
        aria-label={intl.formatMessage({id: 'i18n:studyguides:popup:using-this-guide:close:aria-label'})}
        data-testid='close-utg'
        data-analytics-disable-track={true}
      >
      <CloseIcon/>
    </CloseIconButton>
  </BannerWrapper>;
};

export default UsingThisGuideBanner;
