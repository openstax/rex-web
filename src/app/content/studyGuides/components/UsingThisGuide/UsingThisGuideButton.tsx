import React from 'react';
import { useIntl } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { QuestionCircle } from 'styled-icons/fa-regular/QuestionCircle';
import { PlainButton } from '../../../../components/Button';
import theme from '../../../../theme';
import { toolbarDefaultText } from '../../../components/Toolbar/styled';
import { disablePrint } from '../../../components/utils/disablePrint';
import { mobilePaddingSides } from '../../../styles/PopupConstants';

const buttonMarginTopDesktop = 0.6;
const buttonPaddingTopDesktop = 1.2;
const buttonPaddingBottomDesktop = 1.8;

// tslint:disable-next-line:variable-name
const QuestionIcon = styled(QuestionCircle)`
  height: 1.7rem;
  width: 1.7rem;
`;

// tslint:disable-next-line:variable-name
const UsingThisGuideButtonInnerStyles = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  overflow: visible;
  color: ${theme.color.primary.gray.base};
  position: relative;
  padding: ${buttonPaddingTopDesktop}rem ${mobilePaddingSides}rem ${buttonPaddingBottomDesktop}rem;
  outline: none;
  ${({isOpen}) => isOpen && css`
    color: ${theme.color.white};
  `}
  ${theme.breakpoints.mobile(css`
    padding-right: 1.4rem;
    padding-left: 1.4rem;
  `)}
  ${disablePrint}
`;

// tslint:disable-next-line:variable-name
export const UsingThisGuideButtonWrapper = styled(PlainButton)`
  position: relative;
  height: 100%;
  margin-right: 3.2rem;
  margin-top: ${buttonMarginTopDesktop}rem;
  ${({isOpen}) => isOpen && css`
    background: ${theme.color.black};
  `}
  ${theme.breakpoints.mobile(css`
    margin-right: 0.8rem;
  `)}
`;

// tslint:disable-next-line:variable-name
const UsingThisGuideText = styled.span`
  ${toolbarDefaultText}
`;

interface Props {
  onClick: () => void;
  open: boolean;
}

// tslint:disable-next-line:variable-name
const UsingThisGuideButton = (props: Props) => {
  const text = useIntl().formatMessage({id: 'i18n:studyguides:popup:using-this-guide'});

  return <UsingThisGuideButtonWrapper
    aria-label={text}
    onClick={props.onClick}
    isOpen={props.open}
    data-analytics-disable-track={true}
  >
    <UsingThisGuideButtonInnerStyles isOpen={props.open} tabIndex={-1}>
      <QuestionIcon/>
      <UsingThisGuideText>{text}</UsingThisGuideText>
    </UsingThisGuideButtonInnerStyles>
  </UsingThisGuideButtonWrapper>;
};

export default UsingThisGuideButton;
