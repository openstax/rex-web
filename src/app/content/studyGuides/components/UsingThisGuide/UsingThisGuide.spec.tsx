
import React from 'react';
import renderer from 'react-test-renderer';
import TestContainer from '../../../../../test/TestContainer';
import UsingThisGuideBanner from './UsingThisGuideBanner';
import UsingThisGuideButton from './UsingThisGuideButton';

describe('Using this guide', () => {
  const onclickFn = jest.fn();

  it('renders using this guide button correctly (when banner closed)', () => {
    const component = renderer.create(<TestContainer>
      <UsingThisGuideButton open={false} onClick={onclickFn}/>
    </TestContainer>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders using this guide button correctly (when banner open)', () => {
    const component = renderer.create(<TestContainer>
      <UsingThisGuideButton open={true} onClick={onclickFn}/>
    </TestContainer>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders using this guide banner correctly', () => {
    const component = renderer.create(<TestContainer>
      <UsingThisGuideBanner show={true} onClick={onclickFn}/>
    </TestContainer>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
