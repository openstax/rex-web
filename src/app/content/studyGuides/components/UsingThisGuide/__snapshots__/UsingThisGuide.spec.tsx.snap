// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using this guide renders using this guide banner correctly 1`] = `
<div
  className="UsingThisGuideBanner__BannerWrapper-adj8ti-0 fqaMxX"
>
  <div
    className="UsingThisGuideBanner__HeaderWrapper-adj8ti-5 fwRgqU"
  >
    <h2
      className="headings__H2-sc-1xfiggv-1 UsingThisGuideBanner__UsingThisGuideTitle-adj8ti-2 cFNEOU"
    >
      Using This Guide
    </h2>
  </div>
  <div
    className="UsingThisGuideBanner__BodyWrapper-adj8ti-6 gienZp"
  >
    <picture>
      <source
        media="(max-width: 50em)"
        srcSet="banner_mobile.png"
      />
      <img
        alt="The image presents the structure of study guides. The first element of each block summarizes, clarifies, or enriches key content. The second element is taken directly from the book. On the left side, there is a color label depending on the content."
        className="UsingThisGuideBanner__BannerImage-adj8ti-1 lebSIh"
        src="banner.png"
        tabIndex={0}
      />
    </picture>
  </div>
  <button
    aria-label="Closes the image describing the structure of study guides."
    className="Button__PlainButton-ayg7nk-2 UsingThisGuideBanner__CloseIconButton-adj8ti-3 bnGaoD"
    data-analytics-disable-track={true}
    data-testid="close-utg"
    onClick={[MockFunction]}
  >
    <svg
      aria-hidden="true"
      className="UsingThisGuideBanner__CloseIcon-adj8ti-4 jtVjLX StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 352 512"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`Using this guide renders using this guide button correctly (when banner closed) 1`] = `
<button
  aria-label="Using This Guide"
  className="Button__PlainButton-ayg7nk-2 UsingThisGuideButton__UsingThisGuideButtonWrapper-bu0xmw-2 hHwOGT"
  data-analytics-disable-track={true}
  onClick={[MockFunction]}
>
  <div
    className="UsingThisGuideButton__UsingThisGuideButtonInnerStyles-bu0xmw-1 iGMoSS"
    tabIndex={-1}
  >
    <svg
      aria-hidden="true"
      className="UsingThisGuideButton__QuestionIcon-bu0xmw-0 hHPBbh StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
        fill="currentColor"
      />
    </svg>
    <span
      className="UsingThisGuideButton__UsingThisGuideText-bu0xmw-3 dKabOj"
    >
      Using This Guide
    </span>
  </div>
</button>
`;

exports[`Using this guide renders using this guide button correctly (when banner open) 1`] = `
<button
  aria-label="Using This Guide"
  className="Button__PlainButton-ayg7nk-2 UsingThisGuideButton__UsingThisGuideButtonWrapper-bu0xmw-2 gdZrfW"
  data-analytics-disable-track={true}
  onClick={[MockFunction]}
>
  <div
    className="UsingThisGuideButton__UsingThisGuideButtonInnerStyles-bu0xmw-1 doDlsO"
    tabIndex={-1}
  >
    <svg
      aria-hidden="true"
      className="UsingThisGuideButton__QuestionIcon-bu0xmw-0 hHPBbh StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
        fill="currentColor"
      />
    </svg>
    <span
      className="UsingThisGuideButton__UsingThisGuideText-bu0xmw-3 dKabOj"
    >
      Using This Guide
    </span>
  </div>
</button>
`;
